core/src/main/java/ai/timefold/solver/core/impl/solver/termination/DiminishedReturnsScoreRingBuffer.java

```diff
@@ -0,0 +1,224 @@
+            if (!(o instanceof RingBufferState that)) {
+                return false;
+            }
+            return readIndex == that.readIndex && writeIndex == that.writeIndex && Objects.deepEquals(nanoTimeRingBuffer,
+                    that.nanoTimeRingBuffer) && Objects.deepEquals(scoreRingBuffer, that.scoreRingBuffer);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(readIndex, writeIndex, Arrays.hashCode(nanoTimeRingBuffer), Arrays.hashCode(scoreRingBuffer));
+        }
+
+        @Override
+        public String toString() {
+            return "RingBufferState{" +
+                    "readIndex=" + readIndex +
+                    ", writeIndex=" + writeIndex +
+                    ", nanoTimeRingBuffer=" + Arrays.toString(nanoTimeRingBuffer) +
+                    ", scoreRingBuffer=" + Arrays.toString(scoreRingBuffer) +
+                    '}';
+        }
+    }
+
+    @NonNull
+    RingBufferState getState() {
+        return new RingBufferState(readIndex, writeIndex, nanoTimeRingBuffer, scoreRingBuffer);
+    }
+
+    void resize() {
+        var newCapacity = nanoTimeRingBuffer.length * 2;
+        var newNanoTimeRingBuffer = new long[newCapacity];
+        @SuppressWarnings("unchecked")
+        var newScoreRingBuffer = (Score_[]) new Score[newCapacity];
+
+        if (readIndex < writeIndex) {
+            // entries are [startIndex, writeIndex)
+            var newLength = writeIndex - readIndex;
+            System.arraycopy(nanoTimeRingBuffer, readIndex, newNanoTimeRingBuffer, 0, newLength);
+            System.arraycopy(scoreRingBuffer, readIndex, newScoreRingBuffer, 0, newLength);
+            readIndex = 0;
+            writeIndex = newLength;
+        } else {
+            // entries are [readIndex, CAPACITY) + [0, writeIndex)
+            var firstLength = nanoTimeRingBuffer.length - readIndex;
+            var secondLength = writeIndex;
+            var totalLength = firstLength + secondLength;
+
+            System.arraycopy(nanoTimeRingBuffer, readIndex, newNanoTimeRingBuffer, 0, firstLength);
+            System.arraycopy(scoreRingBuffer, readIndex, newScoreRingBuffer, 0, firstLength);
+            System.arraycopy(nanoTimeRingBuffer, 0, newNanoTimeRingBuffer, firstLength, secondLength);
+            System.arraycopy(scoreRingBuffer, 0, newScoreRingBuffer, firstLength, secondLength);
+            readIndex = 0;
+            writeIndex = totalLength;
+        }
+        nanoTimeRingBuffer = newNanoTimeRingBuffer;
+        scoreRingBuffer = newScoreRingBuffer;
+    }
+
+    public void clear() {
+        readIndex = 0;
+        writeIndex = 0;
+        Arrays.fill(nanoTimeRingBuffer, 0);
+        Arrays.fill(scoreRingBuffer, null);
+    }
+
+    /**
+     * Returns the first element of the score ring buffer.
+     * Does not remove the element.
+     *
+     * @return the first element of the score ring buffer
+     */
+    public @NonNull Score_ peekFirst() {
```

Comment:
Since the array starts as empty, this should not be annotated as `@NonNull`, because `null` is entirely possible.

Reply:
The buffer starts with the score at the start of a phase, and polling will always keep one element in the buffer. Thus, it is an impossible state to have the buffer empty when it's called.